File: 動的計画法/96_Unique Binary Search Trees.js
// 96_Unique Binary Search Trees
// 求n => 考虑规模为i的问题, opt[i] = opt[k] * opt[i - k] { for all possible cut k }

function numTrees(n) {
  if (n === 0 || n === 1) {
    return 1;
  }
  const dp = [1, 1];
  for (let i = 2; i <= n; ++i) {
    dp[i] = 0;
    for (let k = 1; k <= i; ++k) {
      dp[i] = dp[i] + dp[k - 1] * dp[i - k];
    }
  }
  return dp[n];
}

File: 木/96_Unique Binary Search Trees.py
# 96_Unique Binary Search Trees
# DP: dp[n]? 除去根, dp[左] * dp[右], 左: 0 -> n - 1

class Solution:
  def numTrees(self, n):
    dp = [0 for _ in range(n + 1)]
    dp[0] = 1
    dp[1] = 1
    if n < 2:
      return dp[n]
    for i in range(2, n + 1):
      for left in range(i):
        dp[i] += dp[left] * dp[i - left - 1]
    return dp[n]
